File: libs/jgdsl/src/gdsl/asm/Instruction.java
package gdsl.asm;

import gdsl.asm.annotation.Annotation;
import gdsl.asm.operand.Operand;

public class Instruction {
  private long length;

  public long getLength () {
    return length;
  }

  private String mnemonic;

  public String getMnemonic () {
    return mnemonic;
  }

  private Annotation[] annotations;

  public Annotation[] getAnnotations () {
    return annotations;
  }

  private Operand[] operands;

  public Operand[] getOperands () {
    return operands;
  }

  public Instruction (long length, String mnemonic, Annotation[] annotations, Operand[] operands) {
    super();
    this.length = length;
    this.mnemonic = mnemonic;
    this.annotations = annotations;
    this.operands = operands;
  }
}

File: libs/cppgdsl/src/assembly/instruction.cpp
#include "cppgdsl/assembly/instruction.h"
#include "cppgdsl/compare.h"

namespace gdsl {
namespace assembly {

bool instruction::operator==(instruction const& o) const {
  return this->length == o.length && this->mnemonic == o.mnemonic &&
         equals(this->annotations, o.annotations) &&
         equals(this->operands, o.operands);
}

std::ostream& operator<<(std::ostream& out, const instruction& _this) {
  out << _this.get_mnemonic() << " (" << _this.get_length() << "b) ";
  bool first = true;
  for (auto const& opnd : _this.get_operands()) {
    if (first)
      first = false;
    else
      out << " ";
    out << opnd;
  }

  auto const& anns = _this.get_annotations();
  if (anns.size() > 0) {
    out << " [";
    first = true;
    for (auto const& ann : anns) {
      if (first)
        first = false;
      else
        out << ", ";
      out << ann;
    }
    out << "]";
  }

  return out;
}

}  // namespace assembly
}  // namespace gds
